python/love/producer/love_producer_script_queue.py

```diff
@@ -0,0 +1,745 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["LoveProducerScriptQueue"]
+
+import asyncio
+import logging
+
+from collections import deque
+from datetime import datetime
+
+from typing import Dict, Optional, AsyncIterator, Any
+
+from lsst.ts.idl.enums import ScriptQueue, Script
+from lsst.ts.utils import make_done_future
+from lsst.ts.salobj import Domain, Remote, AckError
+from lsst.ts.salobj.base_script import HEARTBEAT_INTERVAL as SCRIPT_HEARTBEAT_INTERVAL
+
+from . import LoveProducerCSC
+
+
+class LoveProducerScriptQueue(LoveProducerCSC):
+    """Specialized LOVE producer to deal with the ScriptQueue CSC."""
+
+    cmd_timeout = 5.0  # command timeout in seconds.
+    max_script_log_messages = 20
+
+    def __init__(
+        self, domain: Domain, log: Optional[logging.Logger] = None, **kwargs
+    ) -> None:
+
+        if "salindex" not in kwargs:
+            raise RuntimeError(
+                "ScriptQueue is an indexed component, `salindex` must be specified."
+            )
+
+        kwargs_reformatted = kwargs.copy()
+        if "csc" in kwargs_reformatted:
+            kwargs_reformatted.pop("csc")
+
+        super().__init__(
+            domain=domain,
+            csc="ScriptQueue",
+            log=log,
+            remote_readyonly=False,
+            **kwargs_reformatted,
+        )
+
+        self.remote_scripts = Remote(
+            domain,
+            "Script",
+            index=0,
+            readonly=True,
+        )
+
+        self.state = dict(
+            available_scripts=[],
+            enabled=False,
+            running=False,
+            waitingIndices=[],
+            currentIndex=0,
+            finishedIndices=[],
+        )
+
+        self.available_scripts = dict(
+            standard=dict(),
+            external=dict(),
+        )
+
+        self.scripts = dict()
+
+        self.scripts_heartbeat = dict()
+
+        self.scripts_log_messages = dict()
+
+        self.script_reply = dict(evt_logMessage=self.send_script_log_message)
+
+        self._set_revcode_mapping(
+            self.remote_scripts.evt_logMessage.rev_code, "script_logMessage"
+        )
+
+        self._set_template_manager_message(
+            topic_name="script_logMessage",
+            topic=self.remote_scripts.evt_logMessage,
+        )
+
+        self._script_heartbeat_producer_task = None
+
+        self.scripts_schema_task: asyncio.Future = make_done_future()
```

Comment:
How does this work?

Reply:
Ah! good question. So this is to avoid querying for scripts schema more than once at a time. 

Basically, when the producer is dealing with the schema, this task will be executing. If the condition to deal with the schema happens while it is already dealing with schema, it will see that the task is executing and just ignore it.

Reply:
Thanks for the explanation!